File: sky_detection/sky_detector/src/sky_detector_ros.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import matplotlib.pyplot as plt
import cv2
from scipy.signal import medfilt
from scipy import ndimage
import numpy as np
from matplotlib import pyplot as plt


def cal_skyline(mask):
    h, w = mask.shape
    for i in range(w):
        raw = mask[:, i]
        after_median = medfilt(raw, 19)
        try:
            first_zero_index = np.where(after_median == 0)[0][0]
            first_one_index = np.where(after_median == 1)[0][0]
            if first_zero_index > 20:
                mask[first_one_index:first_zero_index, i] = 1
                mask[first_zero_index:, i] = 0
                mask[:first_one_index, i] = 0
        except:
            continue
    return mask


def get_sky_region_gradient(img):

    h, w, _ = img.shape

    img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    img_gray = cv2.blur(img_gray, (9, 3))
    cv2.medianBlur(img_gray, 5)
    lap = cv2.Laplacian(img_gray, cv2.CV_8U)
    gradient_mask = (lap < 6).astype(np.uint8)

    kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (9, 3))

    mask = cv2.morphologyEx(gradient_mask, cv2.MORPH_ERODE, kernel)
    # plt.imshow(mask)
    # plt.show()
    mask = cal_skyline(mask)
    after_img = cv2.bitwise_and(img, img, mask=mask)

    return after_img



def image_callback(data):
    try:
        # Convert ROS Image message to OpenCV format
        cv_image = bridge.imgmsg_to_cv2(data, desired_encoding="bgr8")
        # Process the image
        img_sky = get_sky_region_gradient(cv_image)
        # Convert the processed image back to ROS Image message
        img_sky_rosmsg = bridge.cv2_to_imgmsg(img_sky, encoding="bgr8")
        # Publish the processed image
        result_publisher.publish(img_sky_rosmsg)
        # Show the processed image (optional)
        # plt.imshow(img_sky, cmap='gray')
        # plt.title('Processed Image')
        # plt.axis('off')
        # plt.show(block=False)  # Display the image without blocking
    except Exception as e:
        rospy.logerr("Error processing image: {}".format(e))



if __name__ == '__main__':

    rospy.init_node('sky_detector_ros_py', anonymous=True)
    
    bridge = CvBridge()
    # Define the publisher for the processed image
    result_publisher = rospy.Publisher('/processed_image', Image, queue_size=10)
    # Subscribe to the image topic
    rospy.Subscriber('/camera/fisheye1/image_raw', Image, image_callback)
    # Spin
    rospy.spin()

File: sky_detection/sky_detector/src/sky_detector_ros.cpp
#include <opencv2/opencv.hpp>

#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>


cv::Mat cal_skyline(cv::Mat mask) {
    int h = mask.rows;
    int w = mask.cols;

    for (int i = 0; i < w; ++i) {
        cv::Mat raw = mask.col(i);
        cv::Mat after_median;
        cv::medianBlur(raw, after_median, 19);

        double minVal, maxVal; // Variables to store the minimum and maximum values
        cv::Point minLoc, maxLoc; // Variables to store the locations of the minimum and maximum values
        cv::minMaxLoc(after_median, &minVal, &maxVal, &minLoc, &maxLoc); // Find the minimum and maximum values and their locations

        try {
            int first_zero_index = minLoc.y; // Use the location of the minimum value as the first zero index
            int first_one_index = maxLoc.y; // Use the location of the maximum value as the first one index

            if (first_zero_index > 20) {
                mask.rowRange(first_one_index, first_zero_index).setTo(1);
                mask.rowRange(first_zero_index, h).setTo(0);
                mask.rowRange(0, first_one_index).setTo(0);
            }
        } catch (cv::Exception& e) {
            continue;
        }
    }
    return mask;
}


cv::Mat get_sky_region_gradient(cv::Mat img) {
    cv::Mat img_gray, lap, gradient_mask, mask;

    cv::cvtColor(img, img_gray, cv::COLOR_BGR2GRAY);
    cv::blur(img_gray, img_gray, cv::Size(9, 3));
    cv::medianBlur(img_gray, img_gray, 5);
    cv::Laplacian(img_gray, lap, CV_8U);
    cv::threshold(lap, gradient_mask, 6, 255, cv::THRESH_BINARY_INV);

    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(9, 3));
    cv::morphologyEx(gradient_mask, mask, cv::MORPH_ERODE, kernel);

    mask = cal_skyline(mask);

    cv::Mat after_img;
    img.copyTo(after_img, mask);

    return after_img;
}


void imageCallback(const sensor_msgs::Image::ConstPtr& msg) {
    // Convert ROS image message to OpenCV image
    cv_bridge::CvImagePtr cv_ptr;
    try {
        cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
    } catch (cv_bridge::Exception& e) {
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
    }

    // Process the image
    cv::Mat img = cv_ptr->image;

}

int main(int argc, char** argv) {
    ros::init(argc, argv, "image_processing_node");
    ros::NodeHandle nh;

    // Subscribe to the image topic
    ros::Subscriber sub = nh.subscribe<sensor_msgs::Image>("image_topic", 1, imageCallback);

    ros::spin();

    return 0;
}

